File: app/src/main/java/com/lumiyaviewer/lumiya/react/Subscribable.java
package com.lumiyaviewer.lumiya.react;

import com.lumiyaviewer.lumiya.react.Subscription;
import java.util.concurrent.Executor;
import javax.annotation.Nonnull;
import javax.annotation.Nullable;

public interface Subscribable<K, T> {
    Subscription<K, T> subscribe(@Nonnull K k, @Nonnull Subscription.OnData<T> onData);

    Subscription<K, T> subscribe(@Nonnull K k, @Nonnull Subscription.OnData<T> onData, @Nullable Subscription.OnError onError);

    Subscription<K, T> subscribe(@Nonnull K k, @Nullable Executor executor, @Nonnull Subscription.OnData<T> onData);

    Subscription<K, T> subscribe(@Nonnull K k, @Nullable Executor executor, @Nonnull Subscription.OnData<T> onData, @Nullable Subscription.OnError onError);
}

File: app/src/main/java/com/lumiyaviewer/lumiya/react/Subscribable.kt
package com.lumiyaviewer.lumiya.react

import java.util.concurrent.Executor

interface Subscribable<K, T> {
    fun subscribe(key: K, onData: Subscription.OnData<T>): Subscription<K, T>
    fun subscribe(key: K, onData: Subscription.OnData<T>, onError: Subscription.OnError?): Subscription<K, T>
    fun subscribe(key: K, executor: Executor?, onData: Subscription.OnData<T>): Subscription<K, T>
    fun subscribe(key: K, executor: Executor?, onData: Subscription.OnData<T>, onError: Subscription.OnError?): Subscription<K, T>
}
